src/connection_pool.rs

```diff
@@ -537,6 +535,30 @@ where
     }
 }
 
+impl<C, R, F> Deref for PoolConnection<C, R, F>
+where
+    C: Connection,
+    R: Resolver,
+    F: FnMut(&Backend) -> C + Send
+{
+    type Target = C;
+
+    fn deref(&self) -> &C {
+        &(self.connection_pair.0).1.as_ref().unwrap()
```

Comment:
This would crash the app if that happened. I admit this isn't ideal and at some point it might be worth revisiting this to try an d work out a nicer solution. The reason the connection in the `ConnectionKeyPair` is contained in an `Option` is to avoid any tricky ownership issues when replacing the connection in the pool after it is released. In practice it should be the case that it's never possible that a `PoolConnection` to be dereferenced when the second member of its `ConnectionKeyPair` is not a `Some`. The `claim` function will only ever return success when wrapping a claimed connection in a `Some` as seen [here](https://github.com/joyent/rust-cueball/blob/connection-interface-update/src/connection_pool.rs#L353-L356). The connection `Option` is only ever manipulated when the connection is dropped out of scope [here](https://github.com/joyent/rust-cueball/blob/connection-interface-update/src/connection_pool.rs#L518-L521).

I don't feel quite as bad about this too because in looking at the [source](https://docs.rs/r2d2/0.8.3/src/r2d2/lib.rs.html#576) for the r2d2 connection pool they essentially are doing the same thing.

Reply:
Thanks, that helps a lot!